yt_dlp/extractor/echo360.py

```diff
@@ -0,0 +1,118 @@
+import re
+
+from .common import InfoExtractor
+from ..utils import (
+    determine_ext,
+    float_or_none,
+    traverse_obj,
+    variadic,
+)
+
+
+class Echo360IE(InfoExtractor):
+    _INSTANCES_RE = r'''(?:
+                            echo360\.ca|
+                            echo360\.net\.au|
+                            echo360\.org\.au|
+                            echo360\.org\.uk|
+                            echo360\.org|
+                        )'''
+    _UUID_RE = r'[\da-fA-F]{8}-[\da-fA-F]{4}-[\da-fA-F]{4}-[\da-fA-F]{4}-[\da-fA-F]{12}'
+    _VALID_URL = rf'''(?x)
+        https?://(?P<host>{_INSTANCES_RE})
+        /media/(?P<id>{_UUID_RE})/public'''
+
+    _API_BASE = 'https://%s/api/ui/echoplayer/public-links/%s/media/%s/player-properties'
+
+    _TESTS = [
+        {
+            'url': 'https://echo360.org.uk/media/1d8392aa-a3e7-4e78-94cf-b6532c27208c/public',
+            'info_dict': {
+                'id': '3c7ae6e0-fa19-432d-aa21-c283b4276f2a',
+                'ext': 'mp4',
+                'title': '3-4 Force + moment + mechanics.mp4',
+                'duration': 4731.888,
+            },
+            'params': {'skip_download': 'm3u8'}
+        },
+        {
+            'url': 'https://echo360.net.au/media/f04960a9-2efc-4b63-87b5-72e629081d15/public',
+            'info_dict': {
+                'id': '6098a147-2d65-40f3-b9e9-a0204afe450c',
+                'ext': 'mp4',
+                'title': 'EXSC634_Online_Workshop_Week_4.mp4',
+                'duration': 6659.72,
+            },
+            'params': {'skip_download': 'm3u8'}
+        },
+    ]
+
+    def _call_api(self, host, video_id, media_id, session_token, **kwargs):
+        return self._download_json(
+            self._API_BASE % (host, video_id, media_id), video_id,
+            headers={'Authorization': f'Bearer {session_token}'}, **kwargs)
+
+    @staticmethod
+    def _update_url_query(uri, query_string):
+        if query_string is not None:
+            return f'{uri.split("?", 1)[0]}?{query_string}'
+        return uri
+
+    @staticmethod
+    def _get_query_string(uri, query_strings):
+        uri_base = uri.split("?", 1)[0]
+        for query_string in query_strings:
+            if re.match(query_string['uriPattern'], uri_base):
```

Comment:
This is weird. The server gives valid Python regex in `uriPattern`? Pls give an example of `query_strings`

Reply:
js regex. I studied the source of echo360 xD they do exactly this. I actually just copied the source over and converted it to python. 

for the first test:
```
query_strings = [{'uriPattern': '^https://.*\\.echo360\\.org\\.uk/0000\\.ed3e4994-e652-4cd0-9aef-9f58b887736b/3c7ae6e0-fa19-432d-aa21-c283b4276f2a/1/.*$', 'queryString': 'x-uid=a0000000-0000-4000-0000-000000000000&x-instid=ed3e4994-e652-4cd0-9aef-9f58b887736b&x-src=desktop&Policy=eyJTdGF0ZW1lbnQiOiBbeyJSZXNvdXJjZSI6Imh0dHBzOi8vKi5lY2hvMzYwLm9yZy51ay8wMDAwLmVkM2U0OTk0LWU2NTItNGNkMC05YWVmLTlmNThiODg3NzM2Yi8zYzdhZTZlMC1mYTE5LTQzMmQtYWEyMS1jMjgzYjQyNzZmMmEvMS8qP3gtdWlkPWEwMDAwMDAwLTAwMDAtNDAwMC0wMDAwLTAwMDAwMDAwMDAwMCZ4LWluc3RpZD1lZDNlNDk5NC1lNjUyLTRjZDAtOWFlZi05ZjU4Yjg4NzczNmImeC1zcmM9ZGVza3RvcCIsIkNvbmRpdGlvbiI6eyJEYXRlTGVzc1RoYW4iOnsiQVdTOkVwb2NoVGltZSI6MTY3ODg4NDI4OH19fV19&Signature=F9H0WommtDM0jZJdpiqcYyeFluIho8EDWzbvagnkjIx9-2aimqIDeetNsiSPz4K4Z6fI%7EOS8yYe%7EbNDdhDhGmwe3M7HrmdyP8prb6OJBkknLuD9GaE8oSElU9piRxtplOPB2-ltLCuDbI507rNqjhofsw5ASN9z7E4u4o6QllHAMKkVk3evMNKF%7ESmbny01LqzkaYmgXudQbYgCwn0mvRKbofxnC4xcKyALr41DbrxtGW3dle%7EgJypox10LaP4h0H4eDEslUvZhkvx-E-m-SH0n5NLZmoeEAKEEUE7LrjK8aQQfORTv7bFYpDmNxbGKNj%7EkE5N17EtkzI7A0JZ8HBg__&Key-Pair-Id=APKAIPMYRDQXV3PXG2XA'}]
```

for the second test:
```
query_strings = [{'uriPattern': '^https://.*\\.echo360\\.net\\.au/0000\\.d88cbb8d-7036-448f-ad95-946005ec9d97/6098a147-2d65-40f3-b9e9-a0204afe450c/1/.*$', 'queryString': 'x-uid=a0000000-0000-4000-0000-000000000000&x-instid=d88cbb8d-7036-448f-ad95-946005ec9d97&x-src=desktop&Policy=eyJTdGF0ZW1lbnQiOiBbeyJSZXNvdXJjZSI6Imh0dHBzOi8vKi5lY2hvMzYwLm5ldC5hdS8wMDAwLmQ4OGNiYjhkLTcwMzYtNDQ4Zi1hZDk1LTk0NjAwNWVjOWQ5Ny82MDk4YTE0Ny0yZDY1LTQwZjMtYjllOS1hMDIwNGFmZTQ1MGMvMS8qP3gtdWlkPWEwMDAwMDAwLTAwMDAtNDAwMC0wMDAwLTAwMDAwMDAwMDAwMCZ4LWluc3RpZD1kODhjYmI4ZC03MDM2LTQ0OGYtYWQ5NS05NDYwMDVlYzlkOTcmeC1zcmM9ZGVza3RvcCIsIkNvbmRpdGlvbiI6eyJEYXRlTGVzc1RoYW4iOnsiQVdTOkVwb2NoVGltZSI6MTY3ODg4NDQyM319fV19&Signature=PDN9w-MVBz5FeaTvei8zDhdjd4Q2AGx1Qm8lWzB1I1KrUz6C3xAgBEiD%7EyW-GshBmRRZT8eoXFRjuefESKraw4APkdE8PHRKxi8RrfGVMFlWKjTNcw%7EtdWZ9aW0cbePm2%7EHB2Qh5Z3rjNjyTSvtBPuhTdWcDsrUcCbWBdUAW7igBOeMM4TExqzQYNWxDKj1kZLcl-E3%7E3UWve5m8MxdKuj9S4Iqs0UuTXwJoh6TbrbJZeEtOWbHGXOL4CN6HPF2s7tjqNXWYU2QmfljttB4NtrCaHT%7E4leznGYT22YaLrToF0R9UkSr-7K0LH4Wy4yzJc1MqtoTY-OOllXWoAcnf4A__&Key-Pair-Id=APKAIPMYRDQXV3PXG2XA'}]
```
It is pretty simple regex, I do not think they will use something only available in js

Reply:
Ok. But catch `re.error` just in case